man/gcin.1
.TH GCIN 1 "2 FEB 2008" "GCIN 1.4.4" "gcin input method platform"
.SH NAME
gcin \- an input method server
.SH SYNOPSIS
.B gcin
.B [\fB-h\fP|\fB-v\fP|\fB--version\fP]
.br
.SH DESCRIPTION
.B gcin
is an input method(IM) server, which focused mainly on Traditional Chinese. It is also very useful for Simplified Chinese, Japanese, and many other languages. IM table format of gcin is almost as same as those of scim and xcin. Users may append their IMs to gcin very easily. The "g" of gcin means that it features a better GTK+ user interface.

If you are looking for configuration tool of gcin, please see the manpage of \fBgcin-tools\fP.
.SH OPTIONS
.TP
.B \-h, \-v, \-\-version
Show version of gcin.
.SH FILES
.IP /usr/share/gcin/table/gtab.list
Input methods available for gcin. You may copy it to $HOME/.gcin/ and edit it as you like. If you have no $HOME/.gcin/gtab.list, /usr/share/gcin/table/gtab.list will be used.
.IP /usr/share/gcin/script/html-browser
A file helps gcin finding html browsers when needed. You may add any compatible html browser to it.
.IP /usr/share/gcin/script/utf8-edit
A file helps gcin finding text editors when needed. You may add any UTF-8 compatible text editor to it.
.SH ENVIRONMENT VARIABLES
.IP LC_CTYPE
For bash: export LC_CTYPE=zh_TW.UTF-8
.br
For csh: setenv LC_CTYPE zh_TW.UTF-8
.IP XMODIFIERS
For bash: export XMODIFIERS=@im=gcin
.br
For csh: setenv XMODIFIERS @im=gcin
.IP GTK_IM_MODULE
For bash: export GTK_IM_MODULE=gcin
.br
For csh: setenv GTK_IM_MODULE gcin
.IP QT_IM_MODULE
For bash: export QT_IM_MODULE=gcin
.br
For csh: setenv QT_IM_MODULE gcin
.IP GCIN_CONNECT_MSG_ON
Display messages for socket connections.
.IP GCIN_DBG_TMP
Display additional debug messages.
.IP GCIN_ERR_COREDUMP
Option for debugging.
.IP GCIN_IM_CLIENT_NO_AUTO_EXEC
Disable auto-evoke in gcin protocol.
.IP GCIN_OUTPUT_FILTER
Assign output filter to process data before the IM server send them to clients.
.IP GCIN_TABLE_DIR
Assign directory for IM tables other then /usr/share/gcin/table.
.SH SEE ALSO
.BR gcin-tools (1),
.BR gcin2tab (1),
.BR phoa2d (1),
.BR juyin-learn (1),
.BR gcin-message (1)
.SH AUTHORS
.IP "Edward Der-Hua Liu"
Wrote gcin and associated tools.
.IP "Friend of Chanjei IM, Malaysia <www.chinesecj.com>"
Author of original Chanjei IM version 5 table.
.IP "Dayi Co., Ltd. <http://www.dayi.com>"
Author of Dayi IM.


Changelog
** UTF-8 encoding **
1.4.3:
- 詞音 Firefox 用 Shift 切到英數仍有 window 跑出來的問題。
- 修正許氏鍵盤無法輸入 ,./ 的問題。
- 修正 gdm 輸入密碼錯誤時 greeter coredump 的問題。
- gtab 輸入法由詞庫自動選字改成與詞音一樣用輸入緩衝區的方式。輸入法如嘸蝦米等按空白自動送出第一字的就不支援，如果想要用，可以在 gcin-setup 設定空白鍵的選項。
- qt4 IM module 的支援，skype 上也可以用 Shift 切換中英了
- gcin-setup help 改成用 coolcd 編輯過的 html, 感謝 coolcd
- gcin 現在支援 su root 的中文輸入，必須配合最新的 libgcin-im-client.so.1.1.0
- 詞音在編輯區是空的時候按 Shift+xxxx 鍵輸入標點符號，回直接送出不會放到編輯區。

1.4.2:
- 詞音空白鍵選項設定成輸入空白時，選字會選錯的問題。
- 詞音使用 CapsLock 切換中英時，大小寫不對的問題。

1.4.1:
- 現在可以用 mouse 在預選詞的視窗點選，主要給小朋友用。
- 中文輸入狀態按 Shfit + abc  改成輸入大寫。
- tray 上面的選單現在可以切換是否要念出發音 & 也可以切換小鍵盤。
- firefox 3.0/gnome-terminal 關閉 coredump 的問題。

1.4.0:
- 按 keypad . 現在可以輸入 '. '了 。感謝jackblackevo的建議。
- ctrl-alt-, 有時候需要按兩次才會出現的問題。
- tray panel 如果放在最上面，小鍵盤位置會蓋掉 tray 的問題。
- 詞音在某些情況 coredump 的 bug。
- 符號視窗按滾輪捲動後會自動關閉的問題。
- Cheng-Wei Lin提供的：gcin-setup 選項可以設定是否要使用 詞音 ctrl-E 編輯模式。內定是關閉。
- gtab 大小寫如果有不同的定義，現在螢幕小鍵盤會顯示不同的 keyname
- 符號輸入法，一般使用的方法為加入 Ctrl-Shift 循環內，進入符號輸入法後螢幕小鍵盤會自動顯示，按任一鍵輸入後，會自動切回內定的輸入法，同時關閉螢幕小鍵盤。直接按 Ctrl-Alt-4(原標點倉頡) 進入符號輸入也可以。感謝 winlin 的協助。
- 使用 CapsLock 輸入英數時，不能輸入 `*? 的問題。感謝 coolcd 的 bug report.
- 詞音內部的英數模式現在在彈出式的視窗模式現在不會跑出 window 了。
- tsa2d32 Bus alignment 的問題(如 sparc)。
- 符號視窗改用 coolcd 提供的 symbol-table
- Alt-Shift ←→↓↑ key 現在也可以定義 key 了，default 就是定義成 ←→↓↑。
- gtab 輸入法按滿送出字後，按空白鍵的問題。
- 倚天26鍵注音鍵盤數字鍵沒辦法輸入的問題。
- `? 反引號輸入符號的方式改成與新注音類似。感謝 coolcd 提供的 table
- 詞音斷詞選字的 bug

1.3.9:
- 改用自己的简←→正體字轉換，這樣 sim2trad 就不會有掉字的情形。感謝 LGJ 的協助。
- gtab 輸入法 coredump的 bug

1.3.8:
- 詞音按 Delete/Backspace 把編輯區的內容清光後，現在會自動把 Window 隱藏起來。
-注音【】現在可用來輸入 全形（）。
- gtab 現在支援快碼輸入法
- 詞音在用 Shift 預選詞後不會斷詞選詞的問題。
- Ctrl-Shift-;   現在可以用來輸入全形'：'，不過只限非 XIM 的情況。
- 詞音在中文狀態 Shift Enter 可以用了。
- 在使用彈出是輸入視窗時，符號輸入視窗顯示的問題。
- 新增點選符號視窗裡的按鈕後自動關閉的選項，內定是開啟，可以用 gcin-setup 關閉。
- winlin 提供的新 icons
- 現在可以設定 gtab/詞音 選字 window 裏面選擇鍵的顏色。
- 修正 gtab 的 bug，現在應該可以定義 63 個 keys (沒有實際測過)
- sim2trad 如果有特殊符號就無法翻譯的問題。
- Cheng-Wei Lin提供的：新增了一個緩衝區編輯模式，使用者可以在詞音輸入法下按下 ctrl+e 切換緩衝區編輯模式；在緩衝區編輯模式下，可以用 h 跟 l 來左右移動，按下 x 則可刪除。 
- 提供小鍵盤，可控制所有X window 的程式，不限於輸入法的部份。在gcin tray icon(圖像) 按滑鼠中鍵。此功能放在 libgcin1.so，有用到才會 load 到 memory。也可以用 gcin-kbm-toggle 這個可執行檔切換。
- Big endian CPU 上可能出現的 bug.
- 使用簡體字輸出時，如果有無法轉換的情形，現在改成輸出原來的而不是忽略。

1.3.7:
- 有選項可以設定直接進入中文輸入的狀態，不過只能用在非 XIM 的情況。
- 1.3.6 tray 後面會空白的 bug
- Add Netscape Navigator 9 to gtk-im. Thanks to swyear

1.3.6:
- 詞音斷詞選詞的 bug，造成無法手動斷詞及嚴重影響正確率。
- 當詞音空白鍵選項為「輸入空白」，則右邊的 keypad 數字區可以用來選擇同音字；如果是選擇同音字，則是用來輸入數字。
- 詞音同音字選擇，中文字不會對齊的問題。
- 修正smartboy (小光光)回報 tsin.src 裏面資料的錯誤。
- tray icon 不見的問題，特別是 icewm，目前的解法會造成 gcin tray icon 晚5 秒鐘出現。

1.3.5:
- gcin 現在輸入注音 gtab輸入法時可以念出正在輸入的字，必須安裝語音資料庫於 /usr/share/gcin/ogg or /usr/local/share/gcin/ogg，還需要 ogg123 這個執行檔。
- gcin-setup po fix from caleb
- 新增在英數狀態是否要使用 alt-shift 片語輸入的選項。
- 詞音如果遇到同音詞，且這些同音詞都還沒用過，不會選第一個，造成不方便。
- 提供 Windows ctrl-,./;'  輸入標點符號的方式，不過 '：' 由於會與 ctrl-shift 衝，所以沒辦法用，目前暫時不想去理，反正替代的方式太多了。定義檔為  phrase-ctrl.table，必須中文輸入狀態才可以使用。
- more i18n fix from caleb
- 由某人提供的  --use_tsin=N 選項，可以把詞音的支援拿掉。
- 由 caleb 提供的 gcin-setup-tab
- 符號視窗不會顯示 gtab 輸入法按鍵 tip 的問題。
- 新的 mozilla 家族
- gtab 輸入法如果有比目前更短的拆法，拆字顯示區的顏色會變成黃色。
- gtab 新增是否要用?萬用字元的選項。
- 詞音空白鍵選項中的「送出編輯區內容」選項改成「輸入空白」。

1.3.4:
- 由於如果使用彈出式(popup) 輸入視窗，其實只有在顯示按鍵的時候需要移動視窗，這樣可以減少gcin在 firefox 一類 mozilla 家族程式的奇怪問題。現在也把popup 改成內定的選擇。用來用去，還是比較習慣不用 popup 的方式，不用一直看 tray 是中/英，比較不會按錯。
- gtab 如果選用不讓錯誤鍵進入(非傳統)，按下 space 時是無效的拆字，現在可以直接輸入下一個字，不用按 Escape 清除。
- 「唯一選擇時自動送出」這個選項在某些時候會造成 coredump 的問題。
- \ 現在可以定義輸入法了。
- gtab 輸入法選擇不允許錯誤鍵進入，按空白鍵出現紅色警告字不存在後，可以按空白鍵清除，也可以繼續按下一個未輸入的鍵，紅色也會清除。
-  詞音在按 CapsLock 輸入英文後有時候會造成前面已經對的詞後來又變錯的。
- FREEBSD 用 GTK_IM_MODULE=gcin 無法輸入英文的問題。
- gcin-setup 新增結束 gcin 的功能，只有從 gcin tray menu 內叫出 gcin-setup 才可以看到這個選項。
- gcin 如果重新執行，在之前執行的firefox 內輸入文字，gcin 的視窗就不會自動拉到最上面的問題。需更新 im-client。
- man page update from caleb.
- 由 winlin 提供的 icon files。
- gtab 如果定義的字串裏面含有 '<'，則會出現底下的訊息且無法顯示。Gtk-WARNING **: Failed to set text from markup due to error
- 純注音(ctrl-alt-3)現在也支援拼音了，所以現在可以用拼音注音查詢查詢其他的輸入法。
- build fix from Funda Wang.
- 由 caleb 提供的 gcin-setup i18n po patch，由於會對某些人造成困擾而且大部份的人不需要，configure 需要 --use_i18n=Y啟用。
- gtab 不允定義 & 的問題。
- 新增 詞音/gtab CapsLock英數狀態大小寫的選項。

1.3.3:
- 詞音新增同音字詞選擇鍵在後的選項，這樣做的好處是當你由左到右看到某個字詞符合後，就不用再回頭看是那一個鍵。
- 詞音Tab 斷詞改成可以在任意地方按，代表詞從這邊開頭，如果在結尾按則代表把最後一個字當開頭。例如輸入「他示意蜘蛛」，我們想要的是「他是一隻豬」，這時直接在結尾按 Tab 一次就可以了，代表把豬當成一個單字的詞。當使用者自己按空白鍵選詞時，會把這個手動斷詞清除。不過很可惜的，在某些程式不能用，例如 gedit，因為 gedit沒有把  Tab key 傳給 gcin 處理。為了解決這個問題，也可以改按 Escape。
- 詞音編輯區如果沒有東西，這時按 escape 鍵 gcin 會傳回不處理，所以 AP 會自己處理 escape 鍵。
- 在選字詞 menu 出現時按 escape 應該只把 menu 關閉，不要斷詞。
- gcin2tab 現在遇到沒有定義 keyname 的鍵會中斷而不是忽略。
- 詞音「試試新的」-> [是世新的]，由於單字詞「的」頻率出現次數過高，會造成錯誤，把較長斷詞的比重調高。
- gtab 選擇頁數號碼不會清除的問題。
- gtab 在只有顯示輸入鍵的情況，輸入視窗會超出螢幕的問題。
- gtab 垂直選擇頁數顯示錯誤。
- 簡易輸入法的問題。
- 嘗試解決 gcin 在 beryl/firefox 上的問題。不確定是否有用，在 mandriva cooker 上似乎可以。
- 新增注音鍵盤「神通」。
- 由某位提供的 ./configure --use_tray=N，不想要 tray 時可以用此選項。
- alsa2 與 im-client 起衝突的問題，造成 flash player coredump。必須 update libgcin-im-client.so。
- 為了解決 flashplayer9 當掉的問題，XIM 的 IMdkit 換成 scim 的，現在不用 unset XMODIFIERS 似乎也不會當了。如果你必須經常重新啟動 gcin 最好還是在 firefox 的 start-script 裏面用 unset XMODIFIERS，因為新的 IMdkit 在 gcin 重新啟動有時候還是會造成正在執行的 firefox coredump。而且這樣可以節省 memory。
- gtab 唯一選擇時自動送出的新選項。例如我們在倉頡裏面定義「benq 明基」，當按到 ben 時，由於此時沒有其他的是符合 'ben*'，這時候就會自動送出。
- []\ 現在也可以用來切換輸入法了。已經把拉丁字母及帶調粵拼放上去。

1.3.2:
- 新增編輯 gtab 輸入法外加字詞的功能，可以在 gcin-setup -> gtab 中執行。也可以自己手動編輯 ~/.gcin/xxx.gtab.append，其中 xxx 是你內定的 .gtab 輸入法，例如使用倉頡的人其 filename 就是 ~/.gcin/cj.gtab.append。gcin 會在按 ctrl-space 時檢查 file 是否有被修改，自動 reload。
- candyz 提供的 gtab 垂直選字詞的方式，以解決如果詞很長，無法完全顯示的問題。
- gtab 新增可以用  page up/down or keypad +- 跳頁。對嘸蝦米等 space 送出第一個且定義超過10個時就可以用到。
- gcin2tab 現在不會刪除重複的定義，如果不喜歡，請刪除 .cin 重複定義的部份。
- 超過 5 個 key 的時候，相同拆字鍵的字不會保留原來順序的問題。
- 現在 freebsd 上應該是不用 mergesort 了。
- 因為發現有不少人沒有設定 locale 為 zh_TW，也沒有設定 GTK_IM_MODULE=gcin，造成 install 1.3.1 後無法使用 gcin。所以把 gtk im module 的 language 適用範圍改回 '*'，省掉這個麻煩。
- 如果使用注音輸入拼音，現在可以 juyin-learn 可以顯示成聲調拼音的注音方式。
- gtab 在顯示片語的時候，如果遇到 http:// ftp:// 會不顯示這一部份，現在也改成至少顯示 18個 bytes。
- 新增設定文字及視窗背景顏色的功能。不過取消還原成系統的 theme顏色 後不會馬上還原，必須 pkill gcin。
- noseeing-inst bugfix from LGJ
- 使用一般輸入及萬用字元(*?) 查詢現在會顯示目前及總共的頁數。

gcin.cpp
#include "gcin.h"
#include "config.h"
#include "gcin-version.h"
#include "gtab.h"
#if UNIX
#include <signal.h>
#endif
#if GCIN_i18n_message
#include <libintl.h>
#endif

#if UNIX
Window root;
#endif
Display *dpy;

int win_xl, win_yl;
int win_x, win_y;   // actual win x/y
int dpy_xl, dpy_yl;
DUAL_XIM_ENTRY xim_arr[1];

extern unich_t *fullchar[];
gboolean win_kbm_inited;

char *half_char_to_full_char(KeySym xkey)
{
  if (xkey < ' ' || xkey > 127)
    return NULL;
  return _(fullchar[xkey-' ']);
}


#if UNIX
static void start_inmd_window()
{
  GtkWidget *win = gtk_window_new (GTK_WINDOW_TOPLEVEL);
  gtk_widget_realize (win);
  GdkWindow *gdkwin0 = gtk_widget_get_window(win);
  xim_arr[0].xim_xwin = GDK_WINDOW_XWINDOW(gdkwin0);
  dbg("xim_xwin %x\n", xim_arr[0].xim_xwin);
}
#endif


#if USE_XIM
char *lc;

static XIMStyle Styles[] = {
#if 1
        XIMPreeditCallbacks|XIMStatusCallbacks,         //OnTheSpot
        XIMPreeditCallbacks|XIMStatusArea,              //OnTheSpot
        XIMPreeditCallbacks|XIMStatusNothing,           //OnTheSpot
#endif
        XIMPreeditPosition|XIMStatusArea,               //OverTheSpot
        XIMPreeditPosition|XIMStatusNothing,            //OverTheSpot
        XIMPreeditPosition|XIMStatusNone,               //OverTheSpot
#if 1
        XIMPreeditArea|XIMStatusArea,                   //OffTheSpot
        XIMPreeditArea|XIMStatusNothing,                //OffTheSpot
        XIMPreeditArea|XIMStatusNone,                   //OffTheSpot
#endif
        XIMPreeditNothing|XIMStatusNothing,             //Root
        XIMPreeditNothing|XIMStatusNone,                //Root
};
static XIMStyles im_styles;

#if 1
static XIMTriggerKey trigger_keys[] = {
        {XK_space, ControlMask, ControlMask},
        {XK_space, ShiftMask, ShiftMask},
        {XK_space, Mod1Mask, Mod1Mask},   // Alt
        {XK_space, Mod4Mask, Mod4Mask},   // Windows
};
#endif

/* Supported Encodings */
static XIMEncoding chEncodings[] = {
        "COMPOUND_TEXT",
        0
};
static XIMEncodings encodings;

int xim_ForwardEventHandler(IMForwardEventStruct *call_data);

XIMS current_ims;
extern void toggle_im_enabled();


int MyTriggerNotifyHandler(IMTriggerNotifyStruct *call_data)
{
//    dbg("MyTriggerNotifyHandler %d %x\n", call_data->key_index, call_data->event_mask);

    if (call_data->flag == 0) { /* on key */
//        db(g("trigger %d\n", call_data->key_index);
        if ((call_data->key_index == 0 && gcin_im_toggle_keys==Control_Space) ||
            (call_data->key_index == 3 && gcin_im_toggle_keys==Shift_Space) ||
            (call_data->key_index == 6 && gcin_im_toggle_keys==Alt_Space) ||
            (call_data->key_index == 9 && gcin_im_toggle_keys==Windows_Space)
            ) {
            toggle_im_enabled();
        }
        return True;
    } else {
        /* never happens */
        return False;
    }
}

#if 0
void switch_IC_index(int index);
#endif
void CreateIC(IMChangeICStruct *call_data);
void DeleteIC(CARD16 icid);
void SetIC(IMChangeICStruct * call_data);
void GetIC(IMChangeICStruct *call_data);
int xim_gcin_FocusIn(IMChangeFocusStruct *call_data);
int xim_gcin_FocusOut(IMChangeFocusStruct *call_data);

int gcin_ProtoHandler(XIMS ims, IMProtocol *call_data)
{
//  dbg("gcin_ProtoHandler %x ims\n", ims);

  current_ims = ims;

  switch (call_data->major_code) {
  case XIM_OPEN:
#define MAX_CONNECT 20000
    {
      IMOpenStruct *pimopen=(IMOpenStruct *)call_data;

      if(pimopen->connect_id > MAX_CONNECT - 1)
        return True;

#if DEBUG && 0
    dbg("open lang %s  connectid:%d\n", pimopen->lang.name, pimopen->connect_id);
#endif
      return True;
    }
  case XIM_CLOSE:
#if DEBUG && 0
    dbg("XIM_CLOSE\n");
#endif
    return True;
  case XIM_CREATE_IC:
#if DEBUG && 0
     dbg("CREATE_IC\n");
#endif
     CreateIC((IMChangeICStruct *)call_data);
     return True;
  case XIM_DESTROY_IC:
     {
       IMChangeICStruct *pimcha=(IMChangeICStruct *)call_data;
#if DEBUG && 0
       dbg("DESTROY_IC %d\n", pimcha->icid);
#endif
       DeleteIC(pimcha->icid);
     }
     return True;
  case XIM_SET_IC_VALUES:
#if DEBUG && 0
     dbg("SET_IC\n");
#endif
     SetIC((IMChangeICStruct *)call_data);
     return True;
  case XIM_GET_IC_VALUES:
#if DEBUG && 0
     dbg("GET_IC\n");
#endif
     GetIC((IMChangeICStruct *)call_data);
     return True;
  case XIM_FORWARD_EVENT:
#if DEBUG && 0
     dbg("XIM_FORWARD_EVENT\n");
#endif
     return xim_ForwardEventHandler((IMForwardEventStruct *)call_data);
  case XIM_SET_IC_FOCUS:
#if DEBUG && 0
     dbg("XIM_SET_IC_FOCUS\n");
#endif
     return xim_gcin_FocusIn((IMChangeFocusStruct *)call_data);
  case XIM_UNSET_IC_FOCUS:
#if DEBUG && 0
     dbg("XIM_UNSET_IC_FOCUS\n");
#endif
     return xim_gcin_FocusOut((IMChangeFocusStruct *)call_data);
  case XIM_RESET_IC:
#if DEBUG && 0
     dbg("XIM_UNSET_IC_FOCUS\n");
#endif
     return True;
  case XIM_TRIGGER_NOTIFY:
#if DEBUG && 0
     dbg("XIM_TRIGGER_NOTIFY\n");
#endif
     MyTriggerNotifyHandler((IMTriggerNotifyStruct *)call_data);
     return True;
  case XIM_PREEDIT_START_REPLY:
#if DEBUG && 1
     dbg("XIM_PREEDIT_START_REPLY\n");
#endif
     return True;
  case XIM_PREEDIT_CARET_REPLY:
#if DEBUG && 1
     dbg("XIM_PREEDIT_CARET_REPLY\n");
#endif
     return True;
  case XIM_STR_CONVERSION_REPLY:
#if DEBUG && 1
     dbg("XIM_STR_CONVERSION_REPLY\n");
#endif
     return True;
  default:
     printf("Unknown major code.\n");
     break;
  }

  return True;
}


void open_xim()
{
  XIMTriggerKeys triggerKeys;

  im_styles.supported_styles = Styles;
  im_styles.count_styles = sizeof(Styles)/sizeof(Styles[0]);

  triggerKeys.count_keys = sizeof(trigger_keys)/sizeof(trigger_keys[0]);
  triggerKeys.keylist = trigger_keys;

  encodings.count_encodings = sizeof(chEncodings)/sizeof(XIMEncoding) - 1;
  encodings.supported_encodings = chEncodings;

  if ((xim_arr[0].xims = IMOpenIM(dpy,
          IMServerWindow,         xim_arr[0].xim_xwin,        //input window
          IMModifiers,            "Xi18n",        //X11R6 protocol
          IMServerName,           xim_arr[0].xim_server_name, //XIM server name
          IMLocale,               lc,
          IMServerTransport,      "X/",      //Comm. protocol
          IMInputStyles,          &im_styles,   //faked styles
          IMEncodingList,         &encodings,
          IMProtocolHandler,      gcin_ProtoHandler,
          IMFilterEventMask,      KeyPressMask|KeyReleaseMask,
          IMOnKeysList, &triggerKeys,
          NULL)) == NULL) {
          p_err("IMOpenIM '%s' failed. Maybe another XIM server is running.\n",
          xim_arr[0].xim_server_name);
  }
}

#endif // if USE_XIM

void load_tsin_db();
void load_tsin_conf(), load_setttings(), load_tab_pho_file();

void disp_hide_tsin_status_row(), gcb_main(), update_win_kbm_inited();
void change_tsin_line_color(), change_win0_style(), change_tsin_color();
void change_win_gtab_style();
void update_item_active_all();
void destroy_inmd_menu();
void load_gtab_list(gboolean);
void change_win1_font();
void set_wselkey(char *s);

static void reload_data()
{
  dbg("reload_data\n");
  load_setttings();
  if (current_method_type()==method_type_TSIN)
    set_wselkey(pho_selkey);

//  load_tsin_db();
  change_win0_style();
  change_win1_font();
  change_win_gtab_style();
//  change_win_pho_style();
  load_tab_pho_file();
  change_tsin_color();
  update_win_kbm_inited();

  destroy_inmd_menu();
  load_gtab_list(TRUE);

  update_item_active_all();
#if USE_GCB
  gcb_main();
#endif
}

void change_tsin_font_size();
void change_gtab_font_size();
void change_pho_font_size();
void change_win_sym_font_size();
void change_win_gtab_style();
extern int win_kbm_on;

static void change_font_size()
{
  load_setttings();
  change_tsin_font_size();
  change_gtab_font_size();
  change_pho_font_size();
  change_win_sym_font_size();
  change_win0_style();
  change_win_gtab_style();
  update_win_kbm_inited();
//  change_win_pho_style();
}

#if UNIX
static int xerror_handler(Display *d, XErrorEvent *eve)
{
  return 0;
}
#endif

#if UNIX
Atom gcin_atom;
#endif
void disp_tray_icon(), toggle_gb_output();

void cb_trad_sim_toggle()
{
  toggle_gb_output();
#if TRAY_ENABLED
  disp_tray_icon();
#endif
}
void execute_message(char *message), show_win_kbm(), hide_win_kbm();
int b_show_win_kbm=0;
void disp_win_kbm_capslock_init();

void kbm_open_close(gboolean b_show)
{
  b_show_win_kbm=b_show;
  if (b_show) {
    show_win_kbm();
    disp_win_kbm_capslock_init();
  } else
    hide_win_kbm();
}

void kbm_toggle()
{
  win_kbm_inited = 1;
  kbm_open_close(!b_show_win_kbm);
}


void reload_tsin_db();
void do_exit();

void message_cb(char *message)
{
//   dbg("message '%s'\n", message);

   if (!strcmp(message, CHANGE_FONT_SIZE)) {
     change_font_size();
   } else
   if (!strcmp(message, GB_OUTPUT_TOGGLE)) {
     cb_trad_sim_toggle();
     update_item_active_all();
   } else
   if (!strcmp(message, KBM_TOGGLE)) {
     kbm_toggle();
   } else
#if UNIX
   if (strstr(message, "#gcin_message")) {
     execute_message(message);
   } else
#endif
   if (!strcmp(message, RELOAD_TSIN_DB)) {
     reload_tsin_db();
   } else
   if (!strcmp(message, GCIN_EXIT_MESSAGE)) {
     do_exit();
   } else
     reload_data();
}

#if UNIX
static GdkFilterReturn my_gdk_filter(GdkXEvent *xevent,
                                     GdkEvent *event,
                                     gpointer data)
{
   XEvent *xeve = (XEvent *)xevent;
#if 0
   dbg("a zzz %d\n", xeve->type);
#endif

   // only very old WM will enter this
   if (xeve->type == FocusIn || xeve->type == FocusOut) {
#if 0
     dbg("focus %s\n", xeve->type == FocusIn ? "in":"out");
#endif
     return GDK_FILTER_REMOVE;
   }

#if USE_XIM
   if (XFilterEvent(xeve, None) == True)
     return GDK_FILTER_REMOVE;
#endif

   return GDK_FILTER_CONTINUE;
}

void init_atom_property()
{
  gcin_atom = get_gcin_atom(dpy);
  XSetSelectionOwner(dpy, gcin_atom, xim_arr[0].xim_xwin, CurrentTime);
}
#endif


void hide_win0();
void destroy_win0();
void destroy_win1();
void destroy_win_gtab();
void free_pho_mem(),free_tsin(),free_all_IC(), free_gtab(), free_phrase(), destroy_tray_win32();

void do_exit()
{
  dbg("----------------- do_ exit ----------------\n");

  free_pho_mem();
  free_tsin();
#if USE_XIM
  free_all_IC();
#endif
  free_gtab();
  free_phrase();

#if 1
  destroy_win0();
  destroy_win1();
  destroy_win_gtab();
#endif

#if WIN32
  destroy_tray_win32();
#endif
  gtk_main_quit();
}

void sig_do_exit(int sig)
{
  do_exit();
}

char *get_gcin_xim_name();
void load_phrase(), init_TableDir();
void init_tray(), exec_setup_scripts();
#if UNIX
void init_gcin_im_serv(Window win);
#else
void init_gcin_im_serv();
#endif
void gcb_main(), init_tray_win32();

#if WIN32
void init_gcin_program_files();
 #pragma comment(linker, "/subsystem:\"windows\" /entry:\"mainCRTStartup\"")
#endif
int win32_tray_disabled = 1;


gboolean delayed_start_cb(gpointer data)
{
#if WIN32
  Sleep(200);
#endif

  win32_tray_disabled = 0;

#if TRAY_ENABLED
  if (gcin_status_tray) {
#if UNIX
    if (gcin_win32_icon)
      init_tray_win32();
    else
      init_tray();
#endif
  }
#endif

  dbg("after init_tray\n");

#if USE_GCB
  if (gcb_position)
    gcb_main();

  dbg("after gcb_main\n");
#endif

  return FALSE;
}

void get_dpy_xyl()
{
	dpy_xl = gdk_screen_width(), dpy_yl = gdk_screen_height();
}

void screen_size_changed(GdkScreen *screen, gpointer user_data)
{
	get_dpy_xyl();
}

#include "lang.h"

int main(int argc, char **argv)
{
#if WIN32
   putenv("PANGO_WIN32_NO_UNISCRIBE=1");
#endif

  gtk_init (&argc, &argv);

#if UNIX
  signal(SIGCHLD, SIG_IGN);
  signal(SIGPIPE, SIG_IGN);

  if (getenv("GCIN_DAEMON")) {
    daemon(1,1);
#if FREEBSD
    setpgid(0, getpid());
#else
    setpgrp();
#endif
  }
#endif

//putenv("GDK_NATIVE_WINDOWS=1");
#if WIN32
  typedef BOOL (WINAPI* pImmDisableIME)(DWORD);
  pImmDisableIME pd;
  HMODULE imm32=LoadLibraryA("imm32");
  if (imm32 && (pd=(pImmDisableIME)GetProcAddress(imm32, "ImmDisableIME"))) {
     (*pd)(0);
  }
  init_gcin_program_files();
  init_gcin_im_serv();
#endif

  set_is_chs();

#if UNIX
  char *lc_ctype = getenv("LC_CTYPE");
  char *lc_all = getenv("LC_ALL");
  char *lang = getenv("LANG");
  if (!lc_ctype && lang)
    lc_ctype = lang;

  if (lc_all)
    lc_ctype = lc_all;

  if (!lc_ctype)
    lc_ctype = "zh_TW.Big5";
  dbg("gcin get env LC_CTYPE=%s  LC_ALL=%s  LANG=%s\n", lc_ctype, lc_all, lang);
#endif

#if USE_XIM
  char *t = strchr(lc_ctype, '.');
  if (t) {
    int len = t - lc_ctype;
#if MAC_OS || FREEBSD
    lc = strdup(lc_ctype);
    lc[len] = 0;
#else
    lc = g_strndup(lc_ctype, len);
#endif
  }
  else
    lc = lc_ctype;

  char *xim_server_name = get_gcin_xim_name();

  strcpy(xim_arr[0].xim_server_name, xim_server_name);

  dbg("gcin XIM will use %s as the default encoding\n", lc_ctype);
#endif

  if (argc == 2 && (!strcmp(argv[1], "-v") || !strcmp(argv[1], "--version") || !strcmp(argv[1], "-h")) ) {
    p_err(" version %s\n", GCIN_VERSION);
  }

  init_TableDir();
  load_setttings();
  load_gtab_list(TRUE);


#if GCIN_i18n_message
  bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
  textdomain(GETTEXT_PACKAGE);
#endif

  dbg("after gtk_init\n");

#if UNIX
  dpy = GDK_DISPLAY();
  root=DefaultRootWindow(dpy);
#endif
  get_dpy_xyl();
  g_signal_connect(gdk_screen_get_default(),"size-changed", G_CALLBACK(screen_size_changed), NULL);

  dbg("display width:%d height:%d\n", dpy_xl, dpy_yl);

#if UNIX
  start_inmd_window();
#endif

#if USE_XIM
  open_xim();
#endif

#if UNIX
  gdk_window_add_filter(NULL, my_gdk_filter, NULL);

  init_atom_property();
  signal(SIGINT, sig_do_exit);
  signal(SIGHUP, sig_do_exit);
  // disable the io handler abort
  // void *olderr =
    XSetErrorHandler((XErrorHandler)xerror_handler);
#endif

#if UNIX
  init_gcin_im_serv(xim_arr[0].xim_xwin);
#endif

  exec_setup_scripts();

#if UNIX
  g_timeout_add(5000, delayed_start_cb, NULL);
#else
  delayed_start_cb(NULL);
#endif

  dbg("before gtk_main\n");

  disp_win_kbm_capslock_init();

  gtk_main();

  return 0;
}
